File: Mac/TFEGameController+Keyboard.m
//
//  TFEGameController+Keyboard.m
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 9/30/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

// Include Carbon header for key code enum
#import <Carbon/Carbon.h>

#import "TFEGameController+Keyboard.h"

#import "TFEGameController+Private.h"
#import "TFEBoard.h"
#import "TFEMainScene.h"
#import "TFEMainScene+SlowForDebug.h"
#import "TFENodeDirection.h"

@implementation TFEGameController (Keyboard)

- (void)keyDown:(NSEvent *)theEvent
{
    if( NSControlKeyMask == ([theEvent modifierFlags] & NSDeviceIndependentModifierFlagsMask) ){
        if( kVK_ANSI_S == [theEvent keyCode] ){
            [[self scene] toggleSlowForDebug];
            return;
        }
    }
    
    // Don't accept input while movement is taking place.
    if( [[self scene] anyMovementInProgress] ){
        return;
    }
    
    unsigned short code = [theEvent keyCode];
    TFENodeDirection direction = TFENodeDirectionNotADirection;
    switch (code) {
        case kVK_ANSI_A:
        case kVK_LeftArrow:
            direction = TFENodeDirectionLeft;
            break;
        case kVK_ANSI_W:
        case kVK_UpArrow:
            direction = TFENodeDirectionUp;
            break;
        case kVK_ANSI_D:
        case kVK_RightArrow:
            direction = TFENodeDirectionRight;
            break;
        case kVK_ANSI_S:
        case kVK_DownArrow:
            direction = TFENodeDirectionDown;
            break;
        default:
            /* No movement for other keys */
            return;
    }
    
    [[self board] moveNodesInDirection:direction];
}

@end

File: Mac/TFEGameController+Keyboard.swift
//
//  TFEGameController+Keyboard.swift
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 10/2/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

import Cocoa
// Include Carbon for key code enum
import Carbon

extension TFEGameController
{
    override func keyDown(with event: NSEvent)
    {
        let keyCode = Int(event.keyCode)
        let modifierKeys = event.modifierFlags.intersection(NSDeviceIndependentModifierFlagsMask)
        
        if NSEventModifierFlags.control == modifierKeys && kVK_ANSI_S == keyCode {
            self.scene.toggleSlowForDebug()
            return
        }
        
        // Don't accept move input while movement is already taking place. Otherwise nodes get lost.
        guard self.scene.canAcceptInput else {
            return
        }
        
        let direction: SlideDirection
        switch keyCode {
            
            case kVK_ANSI_A, kVK_LeftArrow:
                direction = .left;
            case kVK_ANSI_W, kVK_UpArrow:
                direction = .up;
            case kVK_ANSI_D, kVK_RightArrow:
                direction = .right;
            case kVK_ANSI_S, kVK_DownArrow:
                direction = .down;
            default:
                // No movement for other keys
                return;
        }
        
        self.userDidInput(direction)
    }
}
